File: baekjoon/stage16/1932_정수_삼각형.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n int
	fmt.Fscanln(reader, &n)

	values := [][]int{}
	for i := 0; i < n; i++ {
		inputs := make([]int, i+1)
		for j := 0; j < i+1; j++ {
			fmt.Fscanf(reader, "%d ", &inputs[j])
		}
		values = append(values, inputs)
	}
	maxVal := calculate(values, n)
	fmt.Fprintln(writer, maxVal)
}

func calculate(values [][]int, n int) int {
	if n == 1 {
		return values[0][0]
	}
	maxValues := [][]int{}
	maxValues = append(maxValues, []int{values[0][0]})

	var maxVal int

	for i := 1; i < n; i++ {
		val := []int{}
		for j := 0; j < i+1; j++ {
			if j-1 < 0 {
				maxVal = maxValues[i-1][j]
			} else if j > i-1 {
				fmt.Println(j, i-1)
				maxVal = maxValues[i-1][j-1]
			} else {
				maxVal = int(math.Max(float64(maxValues[i-1][j]), float64(maxValues[i-1][j-1])))
			}
			val = append(val, maxVal+values[i][j])
		}
		maxValues = append(maxValues, val)
	}
	return maxOfArray(maxValues[n-1])
}

func maxOfArray(arr []int) (max int) {
	for _, v := range arr {
		if v > max {
			max = v
		}
	}
	return
}

File: problem-solving/baekjoon/step16/1932_정수_삼각형.py
import sys

def calculate(n):
    if n == 1:
        return values[0][0]
    max_values = [[values[0][0]]]
    idx = 0
    for i in range(1, n):
        val = []
        for j in range(0, i+1):
            if j-1<0:
                max_val = max_values[i-1][j]
            elif j>i-1:
                max_val = max_values[i-1][j-1]
            else:
                max_val = max(max_values[i-1][j], max_values[i-1][j-1])
                
            val.append(max_val+values[i][j])
        max_values.append(val)
    return max(max_values[n-1])


if __name__ == "__main__":
    n = int(sys.stdin.readline())
    values = []
    for i in range(n):
        values.append(list(map(int, sys.stdin.readline().split())))
    max_value = calculate(n)
    print(max_value)
